File: Java/ClotheSample/src/main/java/application/delete/DeleteClotheCommandHandler2.java
package application.delete;

import application.results.Result;
import domain.entities.ClotheEntity;
import domain.repositories.IRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestTemplate;
import org.springframework.http.ResponseEntity;

import javax.validation.ConstraintViolation;
import javax.validation.Validator;
import java.util.Optional;
import java.util.Set;
import java.util.UUID;

@Service
public class DeleteClotheCommandHandler2 {

    public static final String API_URL = "http://127.0.0.1/api/allowDelete";

    private final IRepository repository;
    private final RestTemplate restTemplate;
    private final Validator validator;

    @Autowired
    public DeleteClotheCommandHandler2(IRepository repository, RestTemplate restTemplate, Validator validator) {
        this.repository = repository;
        this.restTemplate = restTemplate;
        this.validator = validator;
    }

    public Result handle(DeleteClotheCommand request) {
        // Validate the command
        Set<ConstraintViolation<DeleteClotheCommand>> violations = validator.validate(request);
        if (!violations.isEmpty()) {
            return new Result(false, "Invalid data", null);
        }

        // Fetch the entity from the repository
        Optional<ClotheEntity> clothe = repository.findById(request.getId());
        if (clothe.isEmpty()) {
            return new Result(false, "Clothe not found", null);
        }

        // Make the HTTP request to check if deletion is allowed
        ResponseEntity<Result> response = restTemplate.getForEntity(API_URL, Result.class);

        if (!response.getStatusCode().is2xxSuccessful()) {
            return new Result(false, "Failed to call external API", null);
        }

        // Process the API response
        Result apiData = response.getBody();
        if (apiData == null) {
            throw new IllegalStateException("Invalid API data received.");
        }

        if (apiData.isSuccess()) {
            // Perform the deletion
            repository.delete(clothe.get());
            return new Result(true, "Clothe deleted successfully", clothe.get().getId());
        } else {
            return apiData;
        }
    }
}

File: .Net/UnitTest/src/Clothe.Application/DeleteClothe/DeleteClotheCommandHandler2.cs
using MediatR;
using Clothes.Application.Results;
using Clothes.Domain.Entities;
using Clothes.Domain.Repositories;
using System.Net.Http;
using System.Text.Json;

namespace Clothes.Application.DeleteClothe;

public class DeleteClotheCommandHandler2(IRepository<ClotheEntity> _repository, HttpClient _httpClient, DeleteClotheCommandValidator _validator) : IRequestHandler<DeleteClotheCommand, Result>
{
    public  const string ApiUrl = "http://127.0.0.1/api/allowDelete";
    public async Task<Result> Handle(DeleteClotheCommand request, CancellationToken cancellationToken)
    {
        var isValid = await _validator.ValidateAsync(request, cancellationToken);
        if (!isValid.IsValid)
        {
            return new Result(false, "Invalid data", isValid.Errors);
        }
        var Clothe = await _repository.GetByIdAsync(request.Id, cancellationToken);
        if (Clothe is null)
        {
            return new Result(false, "Clothe not found");
        }
        var response = await _httpClient.GetAsync(ApiUrl);
        response.EnsureSuccessStatusCode();
        var responseContent = await response.Content.ReadAsStringAsync();

        // Step 2: Process the API response
        var apiData = JsonSerializer.Deserialize<Result>(responseContent);
        if (apiData == null)
        {
            throw new InvalidOperationException("Invalid API data received.");
        }
        if (apiData.Success)
        {
            _repository.Delete(Clothe);
            await _repository.SaveChangesAsync(cancellationToken);
            return new Result(true, "Clothe deleted successfully", Clothe.Id);
        }
        else
        {
            return apiData;
        }
        
    }
}
